File: LeetCode/python3/979.py
class Solution:
    total = 0

    def distributeCoins(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        self.total = 0
        value = self.CountCoins(root)
        return self.total

    def CountCoins(self, node):
        if not node:
            return 0
        value = node.val - 1 + self.CountCoins(node.left) + self.CountCoins(node.right)
        self.total += abs(value)
        return value

File: LeetCode/cpp/979.cpp
class Solution {
public:
    int distributeCoins(TreeNode *root) {
        if (!root)
            return 0;
        int total = 0;
        CountCoins(root, total);
        return total;
    }

    int CountCoins(TreeNode *node, int &total) {
        if (!node)
            return 0;
        int &&value = node->val - 1 + CountCoins(node->left, total) + CountCoins(node->right, total);
        total += abs(value);
        return value;
    }
};
